lib/pure/strmisc.nim

```diff
@@ -85,6 +86,85 @@ proc rpartition*(s: string, sep: string): (string, string, string)
+  assert sep != '-' and decimalDot notin {'-', ' '} and sep != decimalDot
+
+  proc raiseError(i: int; c: char; s: string) {.noinline, noreturn.} =
+    raise newException(ValueError,
+      "Invalid float containing thousand separators, invalid char $1 at index $2 for input $3" %
+      [c.repr, $i, s])
+
+  var s = newStringOfCap(str.len)
+  var successive: int
+  var afterDot, lastWasDot, lastWasSep, hasAnySep, isNegative: bool
+  for idx, c in str:
+    if c in '0' .. '9':  # Digits
+      if hasAnySep and successive > 2:
+        raiseError(idx, c, $str)
+      else:
+        s.add c
+        lastWasSep = false
+        lastWasDot = false
+        inc successive
+    if c == sep:  # Thousands separator, this is NOT the dot
+      if lastWasSep or afterDot or (isNegative and idx == 1 or idx == 0):
+        raiseError(idx, c, $str)
+      else:
+        lastWasSep = true # Do NOT add the Thousands separator here.
+        hasAnySep = true
+        successive = 0
+    if c == decimalDot:  # This is the dot
+      if (isNegative and idx == 1 or idx == 0) or (hasAnySep and successive != 3):  # Disallow .1
+        raiseError(idx, c, $str)
+      else:
+        s.add '.' # Replace decimalDot to '.' so parseFloat can take it.
+        successive = 0
+        lastWasDot = true
+        afterDot = true
+    if c == '-':  # Allow negative float
+      if isNegative or idx != 0:  # Disallow ---1.0
+        raiseError(idx, c, $str)
+      else:
+        s.add '-'
+        isNegative = true
+  parseFloat(s)
+
+
+func parseFloatThousandSep*(s: string; sep = ','; decimalDot = '.'): float {.since: (1, 5).} =
+  ## Convenience func for `parseFloat` which allows for thousand separators,
+  ## this is designed to parse floats as found in the wild formatted for humans.
+  ##
+  ## The following assumptions and requirements must be met:
+  ## - String must not be empty.
+  ## - String must be stripped of trailing and leading whitespaces.
+  ## - `sep` must not be `'-'`.
+  ## - `decimalDot` must not be `'-'` nor `' '`.
+  ## - `sep` and `decimalDot` must be different.
+  ## - Exactly 1 separator must appear after each 3 consecutive inner digits to the left of the dot and nowhere else.
+  ##
+  ## See also:
+  ## * `parseFloat <strutils.html#parseFloat,string>`_
+  runnableExamples:
+    doAssert parseFloatThousandSep("1,000") == 1000.0
+    doAssert parseFloatThousandSep("-1,000") == -1000.0
+    doAssert parseFloatThousandSep("10,000.000") == 10000.0
+    doAssert parseFloatThousandSep("1,000,000.000") == 1000000.0
+    doAssert parseFloatThousandSep("10,000,000.000") == 10000000.0
+    doAssert parseFloatThousandSep("10.000,0", '.', ',') == 10000.0
+    doAssert parseFloatThousandSep("1'000'000,000", '\'', ',') == 1000000.0
+  if s.len > 1: parseFloatThousandSepImpl(s, sep, decimalDot) else: parseFloat(s)
+
+
+func parseFloatThousandSep*(s: openArray[char]; sep = ','; decimalDot = '.'): float {.since: (1, 5).} =
```

Comment:
why a separate overload? `openArray[char]` accepts string as well; the point of using `openArray[char]` instead of string is it allows client code to use slices eg:

`parseFloatThousandSep(input.toOpenArray(1, 4))`

Reply:
Oh, you mean to remove the `string` one?.

Reply:
yes